chore: let `@typescript-eslint/no-floating-promises` ESLint rule pass

ngbrown opened the pull request:
## Summary of changes

This is just the relevant parts of the code changes from #6325 instead of also tackling eslint-config changes.

- Added [`void` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void) to Promises that are not being awaited. Used `await` on finally retry in test.

## Details

When enabling the typescript-eslint rules contained in ["plugin:@typescript-eslint/recommended-requiring-type-checking"](https://typescript-eslint.io/linting/configs#recommended-requiring-type-checking) I tried to improve code quality by addressing some of the issues raised when enabling the type checked linting rules.

This change specifically resolves issues raised with the typescript-eslint error: [`no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises).

> error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises


Testing Strategy:

> I opened up my windows machine and ran this script (with the additional eslint rules enabled from #6325):
>
> ```
> yarn lint
> yarn test:primary
> npx playwrite install
> yarn test
> yarn build --tsc
> ```

The specific rule configuration was:

```js
  "@typescript-eslint/no-floating-promises": [
    "error",
    { ignoreVoid: true, ignoreIIFE: true },
  ],
```

JoshuaKGoldberg commented on packages/remix-dev/cli/commands.ts:
```diff
@@ -208,7 +208,7 @@ export async function watch(
       : await readConfig(remixRootOrConfig);
 
-  devServer.liveReload(config);
+  void devServer.liveReload(config);
```
Out of curiosity, why not `await` here? _(I'm not very familiar with the surrounding code)_

ngbrown replied:
The `await` was moved to the never-resolving `Promise` on the next line in #5223. I'm guessing it's to keep the process running until the `exitHook(() => clean(config))` [in `liveReload.ts`](https://github.com/remix-run/remix/blob/d8bcdc1b209d64cba4769ede49043f59ea22b0dc/packages/remix-dev/devServer/liveReload.ts#L73) has time to finish.

Some of the other floating-promises are in non-async contexts. Others I don't know... is it better to mark the current way it is used or to make them `await` and then possibly face an previously resolved issue?

JoshuaKGoldberg replied:
Cool, thanks for the context. And this is really exciting work btw! 🙌

machour reviewed (commented):
I don't think this needs a changeset since it's internal cosmetics

MichaelDeBoey commented on .changeset/tough-roses-admire.md:
We don't need this changeset, as these changes are all internal cleanup

ngbrown replied:
I removed the changeset.

pcattori commented:
#6325 was closed w/o merging in favor of #6248 , so (I think) this PR is obsolete now. @ngbrown let me know if there's a reason to reopen it.

ngbrown commented:
@pcattori This made code changes that #6248 didn't make, specifically the fix to [withApp.ts#L39](https://github.com/ngbrown-forks/remix/blob/7f6e2f6b343c8231db679655c995c9f1ad9e8825/packages/remix-dev/__tests__/utils/withApp.ts#L39) is a test runtime behavior improvement.

 There were no eslint rule changes in this pull request. I think this pull request should be re-opened.

JoshuaKGoldberg commented:
@MichaelDeBoey just checking in, is there anything else left on this PR?

I'm working on the other recommended typed rules from typescript-eslint and saw this one is still hanging around.

pcattori reviewed (approved):
LGTM, just needs conflicts with `dev` to be resolved

ngbrown commented:
I've rebased the branch, re-ran the [no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises) rule, and updated where there were new errors.

brophdawg11 commented:
Sorry it looks like this has new conflicts 😬 .  I merged dev in locally but don't have access right to push to your fork

ngbrown commented:
@brophdawg11 I re-resolved the conflicts and made sure formatting was applied to `components.tsx` this time.

Because I forked into a separate org, to organize my GitHub forks, I don't have the option to enable maintainers to push to the branch. I didn't realize that was a drawback at the time.
